packages/SwingSet/src/kernel/vat-loader/manager-subprocess-xsnap.js

```diff
@@ -215,19 +227,60 @@ export function makeXsSubprocessFactory({
     mk.setDeliverToWorker(deliverToWorker);
 
     function shutdown() {
+      handleCommandKit?.revoke();
+      handleCommandKit = undefined;
       return worker.close().then(_ => undefined);
     }
     /**
      * @param {number} snapPos
      * @param {SnapStore} snapStore
+     * @param {boolean} [keepWorker]
      * @returns {Promise<SnapshotResult>}
      */
-    function makeSnapshot(snapPos, snapStore) {
-      return snapStore.saveSnapshot(
-        vatID,
+    async function makeSnapshot(snapPos, snapStore, keepWorker) {
+      const snapshotDescription = `${vatID}-${snapPos}`;
+      const snapshotStream = worker.makeSnapshotStream(snapshotDescription);
+
+      if (keepWorker) {
+        return snapStore.saveSnapshot(vatID, snapPos, snapshotStream);
+      }
+
+      /** @type {AsyncGenerator<Uint8Array, void, void>[]} */
+      const [loadStream, saveStream] = synchronizedTee(snapshotStream, 2);
+
+      handleCommandKit?.revoke();
+      handleCommandKit = makeRevokableHandleCommandKit(handleUpstream);
+
+      let snapshotResults;
+      const closeP = worker.close();
+      [worker, snapshotResults] = await Promise.all([
+        startXSnap(argName, {
+          bundleIDs,
+          handleCommand: handleCommandKit.handleCommand,
+          metered,
+          load: {
+            source: 'snapshot',
+            snapshotStream: loadStream,
+            snapshotDescription,
+          },
+        }),
+        snapStore.saveSnapshot(vatID, snapPos, saveStream),
+      ]);
+      await closeP;
+
+      /** @type {Partial<import('@agoric/swing-store/src/snapStore.js').SnapshotInfo>} */
+      const reloadSnapshotInfo = {
         snapPos,
-        worker.makeSnapshotStream(`${vatID}-${snapPos}`),
-      );
+        hash: snapshotResults.hash,
+      };
+
+      kernelSlog.write({
+        type: 'heap-snapshotStream-load',
```

Comment:
I've generally been avoiding camelCase in slog `type:` values, so maybe `heap-snapshot-stream-load`.

Is this slog event in addition to the `save-snapshot` we were producing previously? In that case, if we're distinguishing between just "we saved a snapshot with hash X" and "oh BTW we restarted the worker", then this `type` should look a bit more like the `load-snapshot` event. Maybe `heap-snapshot-reload` ?

Reply:
Oh no I have no idea how this snuck in, looks like a rebase conflict resolution gona bad!

Reply:
We can change this later, but I think the new version will write the following slog items:
* `heap-snapshot-load` (first, beacuse it happens when vatKeeper `saveSnapshot()` calls `manager.makeSnapshot()`)
* `heap-snapshot-save` (second, back in vatKeeper `saveSnapshot()`)

which feels backwards. When we get to a cleanup pass, I'd like to remove the `heap-snapshot-load`, because we now have a `restartWorker` item to the `heap-snapshot-save` slog entry, and I don't want us to confuse this case with a "worker got evicted and now vat-warehouse is bringing it back online" case.

Reply:
Yeah that's a good point.